shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/RouteCardData.kt

```diff
@@ -0,0 +1,349 @@
+        ) : LineOrRoute
+
+        data class Route(val route: com.mbta.tid.mbta_app.model.Route) : LineOrRoute
+    }
+
+    companion object {
+        /**
+         * Build a sorted list of route cards containing realtime data for the given stops.
+         *
+         * Routes are sorted in the following order
+         * 1. pinned routes
+         * 2. subway routes
+         * 3. routes by distance
+         * 4. route pattern sort order
+         *
+         * Any non-typical route patterns which are not happening either at all or between
+         * [filterAtTime] and [filterAtTime] + [hideNonTypicalPatternsBeyondNext] are omitted.
+         * Cancelled trips are also omitted when [context] = NearbyTransit.
+         */
+        fun routeCardsForStopList(
+            stopIds: List<String>,
+            globalData: GlobalResponse?,
+            sortByDistanceFrom: Position?,
+            schedules: ScheduleResponse?,
+            predictions: PredictionsStreamDataResponse?,
+            alerts: AlertsStreamDataResponse?,
+            filterAtTime: Instant,
+            hideNonTypicalPatternsBeyondNext: Duration?,
+            pinnedRoutes: Set<String>,
+            context: Context
+        ): List<RouteCardData>? {
+
+            // if predictions or alerts are still loading, this is the loading state
+            if (predictions == null || alerts == null) return null
+
+            // if global data was still loading, there'd be no nearby data, and null handling is
+            // annoying
+            if (globalData == null) return null
+
+            val cutoffTime = hideNonTypicalPatternsBeyondNext?.let { filterAtTime + it }
+
+            return ListBuilder()
+                .addStaticStopsData(stopIds, globalData)
+                .addUpcomingTrips(schedules, predictions, filterAtTime)
+                .filterIrrelevantData(
+                    cutoffTime,
+                    showAllPatternsWhileLoading = context == Context.StopDetails,
+                    filterCancellations = context == Context.NearbyTransit
+                )
+                .addAlerts(
+                    alerts,
+                    includeMinorAlerts = context == Context.StopDetails,
+                    filterAtTime
+                )
+                .build()
+                .sort(sortByDistanceFrom, pinnedRoutes)
+        }
+    }
+
+    class ListBuilder() {
+        var data: ByLineOrRouteBuilder = emptyMap()
+            private set
+
+        /**
+         * Construct a map of the route/line-ids served by the given stops. Uses the order of the
+         * stops in the given list to determine the stop ids that will be included for each route.
+         *
+         * A stop is only included at a route if it serves any route pattern that is not served by
+         * an earlier stop in the list.
+         */
+        fun addStaticStopsData(stopIds: List<String>, globalData: GlobalResponse): ListBuilder {
```

Comment:
This is mostly a port of [this NearbyStaticData constructor](https://github.com/mbta/mobile_app/blob/7a09634054172e6cb20144c9742dd98d64015d10/shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/NearbyStaticData.kt#L202), building the new data types instead of NearbyStaticData. I did modify the logic a bit to handle routes and lines in the same way rather than as separate steps.